### ISSUE TITLE
Question about dimensions of dimensions of numPost/PrePeriods 

### ISSUE THREAD

**Issue by pachard** [asheshrambachan/HonestDiD#3]
Hello, 

My setting is the following, I observe regions i for three periods, -1, 0 and 1. Treatment happens in certain regions between 0 and 1 (very much like your motivating example). I estimate the following equation, D is a dummy for treated region  

 y_{i,t} = \alpha + \lambda_{t} + \omega_{i} + \beta_{-1}*D*{year=-1} + \beta_{1}*D{year=1} + time varying controls + \eps_{i,t}

where \beta_{-1} is the placebo and \beta_{0} is the omitted/normalized coefficient, this is your equation (1). 

When I create the list (as you did with LWdata_EventSudy), betahat and sigma are of dimension 2 and numPre and Postperiods are both 1.  This is how I understand you dealt with the omitted period. However when I do that, I receive the following error message (when I run createSensitivityResults) 

Error in `[<-`(`*tmp*`, r, r:(r + 2), value = c(1, -2, 1)) : 
  subscript out of bounds

The code does run when I add zeros for \beta_{0} in both the vector of coefficients and the covariance matrix. The dimension of everything goes from two to three (and I change numPrePeriods to 2) but it seems an odd manipulation. 

My questions are: 
- In my case; what should be the dimensions of bethat, sigma and numPre/PostPeriods? 
- How odd is me manually adding zeros? 
- Is there something I got wrong about the way you dealt with the missing coefficient in creating LWdata_EventStudy? 

Thanks again, 
Best, 
Pascal

**Comment by asheshrambachan** [comment]
Just to make sure, sigma should be the full variance-covariance matrix of the data. So, in this example, sigma should be a 2 x 2 matrix. 

I put together a simple toy example based on the description you provided. Here's the code

```
set.seed(1234567890)
betahat = c(mvtnorm::rmvnorm(1, mean = c(0, 0)))
sigma = diag(2)
numPrePeriods = 1 
numPostPeriods = 1
l_vec = basisVector(index = 1, size = 1)
DeltaSD_RobustResults = createSensitivityResults(betahat = betahat,
                                                 sigma = sigma,
                                                 numPrePeriods = numPrePeriods,
                                                 numPostPeriods = numPostPeriods,
                                                 l_vec = l_vec,
                                                 Mvec = c(0, 0.5, 1))
```

In this example, there is 1 pre-period, 1 post-period, betahat is a vector of length 2 and sigma is a 2x2 matrix. Does this match the example you're running?

**Comment by pachard** [comment]
Hello, 

When I run your code, I get the following 

***************************************************************************************************

> set.seed(1234567890) 
> betahat = c(mvtnorm::rmvnorm(1, mean = c(0, 0))) 
> sigma = diag(2) 
> numPrePeriods = 1 
> numPostPeriods = 1 
> l_vec = basisVector(index = 1, size = 1) 
> DeltaSD_RobustResults = createSensitivityResults(betahat = betahat, sigma = sigma, 
+                         numPrePeriods = numPrePeriods, numPostPeriods = numPostPeriods, 
+                         l_vec = l_vec, Mvec = c(0, 0.5, 1))
Warning messages:
1: In seq_len(length.out - 2L) * by :
  Recycling array of length 1 in vector-array arithmetic is deprecated.
  Use c() or as.vector() instead.

2: In seq_len(length.out - 2L) * by :
  Recycling array of length 1 in vector-array arithmetic is deprecated.
  Use c() or as.vector() instead.

3: In seq_len(length.out - 2L) * by :
  Recycling array of length 1 in vector-array arithmetic is deprecated.
  Use c() or as.vector() instead.

> 
> DeltaSD_RobustResults = createSensitivityResults(betahat = betahat, sigma = sigma, 
+                         numPrePeriods = numPrePeriods, numPostPeriods = numPostPeriods, 
+                                                  l_vec = l_vec)
Error in `[<-`(`*tmp*`, r, r:(r + 2), value = c(1, -2, 1)) : 
  subscript out of bounds
> 
> test <- computeConditionalCS_DeltaSD(betahat = betahat, sigma = sigma, numPrePeriods = numPrePeriods,  
+                             numPostPeriods = numPostPeriods, l_vec = l_vec, M=0, 
+                              alpha = 0.05, hybrid_flag = "FLCI", hybrid_kappa = alpha/10,
+                              returnLength = F, postPeriodMomentsOnly = T,
+                              gridPoints =10^3, grid.lb = NA, grid.ub = NA)
Error in rowSums(A_SD[, postPeriodIndices] != 0) : 
  'x' must be an array of at least two dimensions
> 
> test <- computeConditionalCS_DeltaRMI(betahat = betahat, sigma = sigma, numPrePeriods = numPrePeriods,  
+                                       numPostPeriods = numPostPeriods, l_vec = l_vec, Mbar = 0,
+                                       alpha = 0.05, hybrid_flag = "LF", hybrid_kappa = alpha/10,
+                                       returnLength = F, postPeriodMomentsOnly = T,
+                                       gridPoints=10^3, grid.ub = NA, grid.lb = NA)
Error in `[<-`(`*tmp*`, r, (r - 1):r, value = c(1, -1)) : 
  subscript out of bounds 

**************************************************************************************************

so createSensitivityResults only works when Mvec is specified at least when I run it on my computer (in the help documentation it is said that if NULL, the function creates a grid). 

and for  the other two computeConditionalCS_Delta.. I have error results when I use the command as written in the pdf accompanying the package. 

Some of these errors refer to dimensions being lower than two (I guess for pre or post periods). Is it a problem if there are only three periods in my case? 

Thx, 
Pascal

**Comment by asheshrambachan** [comment]
Thanks for these detailed cases! This is incredibly helpful. It appears that all of these errors are due to the special case with one pre-period and one post-period.

The first error on createSensitivityResults when Mvec is not specified arises due to the way the default method the function uses to set Mvec. It does so by constructing an upper-bound for M using the pre-periods but this is only possible if there are > 2 pre-periods (not counting the reference period). The error returned is ``correct'' in the sense that the function cannot construct an upper-bound for M but it specify how to set the default otherwise. I modified the function to set the default Mvec equal to a vector from M = 0 to M = 1*sd of the first post-period coefficient in this case. I pushed a commit that fixes this.

The second two errors arise because of some issues with how the functions are handling the case with just a single post-period (no nuisance parameters). I am closing this issue and opening up a separate issue to fix this problem -- follow the thread over there for updates.
